database/src/service/read.rs

```diff
@@ -0,0 +1,178 @@
+    type Error = oneshot::error::RecvError; // TODO: always unwrap on channel failure?
+    type Future = ResponseRecv;
+
+    #[inline]
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        todo!()
+    }
+
+    #[inline]
+    fn call(&mut self, _req: ReadRequest) -> Self::Future {
+        todo!()
+    }
+}
+
+//---------------------------------------------------------------------------------------------------- DatabaseReader Impl
+/// Database reader thread.
+///
+/// This struct essentially represents a thread.
+///
+/// Each reader thread is spawned with access to this struct (self).
+pub(super) struct DatabaseReader {
+    /// Receiver side of the database request channel.
+    ///
+    /// Any caller can send some requests to this channel.
+    /// They send them alongside another `Response` channel,
+    /// which we will eventually send to.
+    ///
+    /// We (the database reader thread) are not responsible
+    /// for creating this channel, the caller provides it.
+    receiver: crossbeam::channel::Receiver<(ReadRequest, ResponseSend)>,
+
+    /// TODO: either `Arc` or `&'static` after `Box::leak`
+    /// Access to the database.
+    db: &'static ConcreteEnv,
+}
+
+impl DatabaseReader {
+    /// Initialize the `DatabaseReader` thread-pool.
+    ///
+    /// This spawns `N` amount of `DatabaseReader`'s
+    /// attached to `db` and returns a handle to the pool.
+    ///
+    /// Should be called _once_ per actual database.
+    #[cold]
+    #[inline(never)] // Only called once.
+    pub(super) fn init(db: &'static ConcreteEnv) -> DatabaseReadHandle {
+        // Initalize `Request/Response` channels.
+        let (sender, receiver) = crossbeam::channel::unbounded();
+
+        // TODO: slightly _less_ readers per thread may be more ideal.
+        // We could account for the writer count as well such that
+        // readers + writers == total_thread_count
+        let readers = cuprate_helper::thread::threads().get();
+
+        // Spawn pool of readers.
+        for _ in 0..readers {
+            let receiver = receiver.clone();
+
+            std::thread::spawn(move || {
+                let this = Self { receiver, db };
+
+                Self::main(this);
+            });
+        }
+
+        // Return a handle to the pool.
+        DatabaseReadHandle { sender }
+    }
+
+    /// The `DatabaseReader`'s main function.
+    ///
+    /// Each thread just loops in this function.
+    #[cold]
+    #[inline(never)] // Only called once.
+    fn main(mut self) {
+        loop {
+            // 1. Hang on request channel
+            // 2. Map request to some database function
+            // 3. Execute that function, get the result
+            // 4. Return the result via channel
+            let (request, response_send) = match self.receiver.recv() {
+                Ok((r, c)) => (r, c),
+                Err(e) => {
+                    // TODO: what to do with this channel error?
```

Comment:
you just return, an error means all handle are dropped so the DB should shutdown (If we change to and Arc)